src/meta/processors/job/DataBalanceJobExecutor.cpp

```diff
@@ -0,0 +1,197 @@
+    if (LastUpdateTimeMan::update(kvstore_, time::WallClock::fastNowInMilliSec()) !=
+        nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "Balance plan " << plan_->id() << " update meta failed";
+    }
+    executorOnFinished_(status);
+  });
+  plan_->invoke();
+  return Status::OK();
+}
+
+Status DataBalanceJobExecutor::buildBalancePlan() {
+  std::map<std::string, std::vector<Host*>> lostZoneHost;
+  std::map<std::string, std::vector<Host*>> activeSortedHost;
+  for (auto& zoneMapEntry : spaceInfo_.zones_) {
+    for (auto& hostMapEntry : zoneMapEntry.second.hosts_) {
+      activeSortedHost[zoneMapEntry.first].push_back(&hostMapEntry.second);
+    }
+  }
+  for (HostAddr hosts : lostHosts_) {
+    if (!spaceInfo_.hasHost(hosts)) {
+      return Status::Error(
+          "Host %s does not belong to space %d", hosts.toString().c_str(), spaceInfo_.spaceId_);
+    }
+    for (auto& zoneMapEntry : spaceInfo_.zones_) {
+      auto it = zoneMapEntry.second.hosts_.find(hosts);
+      if (it != zoneMapEntry.second.hosts_.end()) {
+        lostZoneHost[zoneMapEntry.first].push_back(&it->second);
+        std::vector<Host*>& hvec = activeSortedHost[zoneMapEntry.first];
+        hvec.erase(std::find(hvec.begin(), hvec.end(), &it->second));
+        break;
+      }
+    }
+  }
+  for (auto& hostMapEntry : activeSortedHost) {
+    std::vector<Host*>& hvec = hostMapEntry.second;
+    std::sort(hvec.begin(), hvec.end(), [](Host*& l, Host*& r) -> bool {
+      return l->parts_.size() < r->parts_.size();
+    });
+  }
+  plan_.reset(new BalancePlan(jobDescription_, kvstore_, adminClient_));
+  // move parts of lost hosts to active hosts in the same zone
+  for (auto& zoneHostEntry : lostZoneHost) {
+    std::vector<Host*>& hvec = activeSortedHost[zoneHostEntry.first];
+    for (Host* host : zoneHostEntry.second) {
+      for (PartitionID partId : host->parts_) {
+        Host* dstHost = hvec.front();
+        dstHost->parts_.insert(partId);
+        plan_->addTask(BalanceTask(jobId_,
+                                   spaceInfo_.spaceId_,
+                                   partId,
+                                   host->host_,
+                                   dstHost->host_,
+                                   kvstore_,
+                                   adminClient_));
+        for (size_t i = 0; i < hvec.size() - 1; i++) {
+          if (hvec[i]->parts_.size() > hvec[i + 1]->parts_.size()) {
+            std::swap(hvec[i], hvec[i + 1]);
+          } else {
+            break;
+          }
+        }
+      }
+      host->parts_.clear();
+    }
+  }
+  lostZoneHost.clear();
+  // rebalance for hosts in a zone
+  auto balanceHostVec = [this](std::vector<Host*>& hostVec) -> std::vector<BalanceTask> {
+    size_t totalPartNum = 0;
+    size_t avgPartNum = 0;
+    for (Host* h : hostVec) {
+      totalPartNum += h->parts_.size();
+    }
+    avgPartNum = totalPartNum / hostVec.size();
```

Comment:
could hostVec be empty?

Reply:
> could hostVec be empty?

 there shouldn't be an empty zone in principle, check if hostVec.size()==0 is more safe, thanks.